### ISSUE TITLE
kargs phase 2

### ISSUE THREAD

**Issue by cgwalters** [openshift/machine-config-operator#798]
Expanding this: https://github.com/openshift/machine-config-operator/pull/762#issuecomment-493453115

We need to support "early pivot" kargs.  I think the best implementation path for this is to "invert" the MachineConfig and inject it as extra data into the Ignition.  Then we change `pivot` to also accept a `MachineConfig` object; it would handle `osImageURL` and `kernelArguments`.

What would make further sense then I think is to fold `pivot` into the MCO repo and have it basically become the "host portion" of the MCD.  But we don't need to do that right away.

**Comment by ashcrow** [comment]
> Then we change pivot to also accept a MachineConfig object; it would handle osImageURL and kernelArguments.

WFM

>  What would make further sense then I think is to fold pivot into the MCO repo and have it basically become the "host portion" of the MCD. But we don't need to do that right away.

I worry that the MCO repo has a lot already (MCO, MCC, MCS, MCD). I won't block but just something to keep in mind.

**Comment by cgwalters** [comment]
> I worry that the MCO repo has a lot already (MCO, MCC, MCS, MCD). I won't block but just something to keep in mind.

Yeah, on that topic one thing I personally would like to move *out* of the MCO is the "static" templates (i.e. files not templated) like [this one](https://github.com/openshift/machine-config-operator/blob/3ec65501fa6e39ed1e4be3b3439138b39aa6fbbe/templates/common/_base/files/cleanup-cni-conf.yaml) - we could reuse the [tools repo](https://github.com/openshift/machine-config-operator/blob/3ec65501fa6e39ed1e4be3b3439138b39aa6fbbe/templates/common/_base/files/cleanup-cni-conf.yaml)?   It's weird for the MCO to feed files into Ignition that could just be shipped as part of the read-only base OS.

**Comment by cgwalters** [comment]
> Yeah, on that topic one thing I personally would like to move out of the MCO is the "static" templates (i.e. files not templated) like this one - we could reuse the tools repo?

Except we can't do that as is because it'd break BYO RHEL...they'd need some other way to add those files.   Aarrgh...

**Comment by cgwalters** [comment]
Eh, if we stick them in an RPM that would work.

**Comment by cgwalters** [comment]
PR in https://github.com/openshift/machine-config-operator/pull/859

**Comment by cgwalters** [comment]
Important PR in https://github.com/openshift/machine-config-operator/pull/868

Another way to look at this is - if we didn't go down this path of having RHCOS boot into MachineConfig (and the MCD on the host etc.) for `kernelArguments` we'd have to do something like "transcode" the MachineConfig kargs into what `pivot` implements today.  We would basically need an ad-hoc way to pass all of the MachineConfig data to pivot.

In fact, the way `osImageURL` [is handled today](https://github.com/openshift/machine-config-operator/blob/c7f5924e73c6d9cf3b5e4a2e3f5fa89752ebf75d/pkg/server/server.go#L53) is very much a weird ad-hoc encoding of a subset of `MachineConfig`.  This path will replace that.

**Comment by cgwalters** [comment]
OK so we're getting deep into this now and I'm about to post a PR which will be a big chunk of converting the firstboot/"early pivot" over to processing MachineConfig.

However...it occurs to me that unless we do some nontrivial tricky work, this will create a hard requirement for users of 4.2 to start from 4.2 *bootimages*.  I've said a lot of times that part of the rationale for the whole current model is that it's hard to rely on users (e.g. hand rolled PXE setups, private clouds) to keep bootimages up to date reliably unless the cluster takes that over.

But can we assume that they'll at least update bootimages for major versions?  I...hope so.  I guess if we do require it we need to provide a clear and unambiguous error if the bootimage is too old, rather than e.g. silently ignoring FIPS early on.

**Comment by cgwalters** [comment]
To be clear, I think we *could* support pivoting from `4.1.0` bootimages, but if e.g. kernel arguments are specified, we'd end up needing to do two reboots. That said, I can think of (somewhat evil) hacks to avoid that like encoding the *new* `machine-config-daemon` code as base64 into Ignition and running it there but...

**Comment by runcom** [comment]
> But can we assume that they'll at least update bootimages for major versions? I...hope so. I guess if we do require it we need to provide a clear and unambiguous error if the bootimage is too old, rather than e.g. silently ignoring FIPS early on.

I'm assuming that as well and it would be really surprised if that won't be the case but (just read the other comment) maybe we can indeed hack something up for 4.1 upgrade paths

**Comment by cgwalters** [comment]
Hm, an important related issue is that we obviously need to support upgrades from 4.1.X to 4.2.  In that case the new MCD will land on the hosts first (thanks to https://github.com/openshift/machine-config-operator/pull/887 ) but the new MCD will have to handle injecting `/etc/pivot/image-pullspec` and using `pivot.service` until it's upgraded.

Probably to be maximally conservative we should keep the code that talks to `pivot` on the host for 4.2.

**Comment by cgwalters** [comment]
I want to cross-link https://github.com/openshift/machine-config-operator/issues/662#issuecomment-506472687 here because I'm increasingly thinking we *need* the reboot-management API as a prerequisite to this if we're going to support using 4.1.X bootimages with 4.2.

And in fact, that's simply going to happen by default for upgrades because today *nothing updates the AMI* in AWS (which we could but that's another story).  Not to mention PXE, private clouds like OpenStack etc.

Ugh...this is making me feel like the "MCD on host" work is...not quite wasted but since we can't *rely* on it it's messy.  Maybe we should just eat the "double reboot" problem for kargs and have the first run of the *cluster* MCD (we're going to need terms here) handle it?

**Comment by runcom** [comment]
> Ugh...this is making me feel like the "MCD on host" work is...not quite wasted but since we can't _rely_ on it it's messy. Maybe we should just eat the "double reboot" problem for kargs and have the first run of the _cluster_ MCD (we're going to need terms here) handle it?

I feel we can afford the double reboot issue for 4.2 at this point? it's just literally a additional reboot if users specify kargs at installation right? next release we can collapse this with rebootless MCO and fix both, how does that sound?
Or maybe we could rely on the old pivot + kargs file for 4.2 to avoid the additional reboot?

**Comment by cgwalters** [comment]
Another alternative is to have the MCS serve Ignition which extracts and runs the MCD from the payload or so.  The messy thing with *that* of course is that the payload image is RHEL7 but the host is RHEL8...so we'd be back in the boat of pulling in e.g. `openssl-deprecated`.  (Which, we added to the host for some storage container that does the same "run container binary on host" but we want to get away from that)

> I feel we can afford the double reboot issue for 4.2 at this point? it's just literally a additional reboot if users specify kargs at installation right?

Yes.  But the core problem noted in https://github.com/openshift/machine-config-operator/issues/662#issuecomment-506472687 is around how we handle that second reboot.  If it's initiated by the payload-MCD once it lands on a host, then *nothing is coordinating those reboots* and you get lots of downtime/disruption.

The other related problem is that the whole MCD model is designed around doing "machineconfig transitions" from config A to config B.  But...in this case (booting from 4.1.0 AMI) then the MCD is going to *think* it's in config A (since that's what the node annotations say) - and it is, except for the kargs.

**Comment by ericavonb** [comment]
> So we'd need to teach the MCD code to either handle the fully general problem of inspecting the host state
I think this is what should be happening. It makes a lot of things a lot more robust in general if the MCD reconciles with the actual host state vs what it thinks the state is.

**Comment by runcom** [comment]
> > So we'd need to teach the MCD code to either handle the fully general problem of inspecting the host state

we're somehow doing that already with currentConfig on-disk so yeah, it sounds right

**Comment by cgwalters** [comment]
> It makes a lot of things a lot more robust in general if the MCD reconciles with the actual host state vs what it thinks the state is.

The underlying rationale behind the current model is here: https://github.com/openshift/machine-config-operator/pull/245

Basically, a long while ago we had *reboot loops* - some other operator was writing a CA cert file on the host that the MCO was also trying to write, and previously we did exactly this "change detection" and...yeah, reboot loops were a *PITA* to debug.

That said, I think that this point most teams know to go through the MCO for changes...maybe we just need to implement a "reboot backoff" or so?  Basically if we're rebooting due to unexpected config drift, then reboot and not long after detect drift again, wait 30 minutes?

Now, that noted, I thought about this a fair bit with kargs.  You can see parts of this debate in the [pivot PR for kargs](https://github.com/openshift/pivot/pull/47) - an issue here is around having lots of code bases parsing e.g. `/proc/cmdline` - including [subtleties](https://github.com/ostreedev/ostree/issues/1859).

**Comment by cgwalters** [comment]
When trying to redesign all of this, I didn't account for the fact that [upgrades won't have newer bootimages](https://github.com/openshift/os/issues/381).  We obviously *must* support upgrades from 4.1 so...here's my new thinking:

First, we keep the work to have MCD baked into the host by default.

But, change the MCS to serve the MCD binary base64 encoded into `/usr/local/bin/machine-config-daemon`, as well as a systemd unit which runs it.  Implementing this is a bit...eww since we need to get that binary from somewhere, and we can't use the one in our own image because that's RHEL7 and not RHEL8.  We can get it out of `machine-os-content` but that's going to be quite involved to scrape.

Anyways, let's call this the "firstboot MCD".  It will run after Ignition and process the [encapsulated MachineConfig(https://github.com/openshift/machine-config-operator/pull/868).  Then the machine reboots, and we continue down the current path of having the "daemonset MCD" exec the on-host MCD where we previously invoked `pivot`.

**Comment by runcom** [comment]
> Anyways, let's call this the "firstboot MCD". It will run after Ignition and process the [encapsulated MachineConfig(https://github.com/[/pull/868](https://github.com/openshift/machine-config-operator/pull/868))

as a side note, I've realized doing this will allow every bits in the MC to have an early pivot way (like fips) by default, anyway...

> But, change the MCS to serve the MCD binary base64 encoded into `/usr/local/bin/machine-config-daemon`, as well as a systemd unit which runs it. Implementing this is a bit...eww since we need to get that binary from somewhere, and we can't use the one in our own image because that's RHEL7 and not RHEL8. We can get it out of `machine-os-content` but that's going to be quite involved to scrape

Couldn't we have a unit which, uhm, extract the MCD from the MCO image in the payload and run it directly (from ignition firstboot)?

**Comment by runcom** [comment]
> ~Couldn't we have a unit which, uhm, extract the MCD from the MCO image in the payload and run it directly (from ignition firstboot)?~ screw this, I just re-read rhel7 vs rhel8 ouch, going `machine-os-content` extract-copy-run would be ok

uhm, grabbing the MCD from the oscontainer is indeed quite tricky (at least for me not being familiar with ostree)

**Comment by cgwalters** [comment]
Probably simplest is to have the MCS mount `machine-os-content` as a volume?  From there, yeah we can extract the binary, though it will require the `ostree` binary in the MCS at least.  This would be easier with https://github.com/openshift/os/issues/382 but that's not happening now.

**Comment by runcom** [comment]
> Probably simplest is to have the MCS mount `machine-os-content` as a volume?

that requires podman also right? (we ship that)

**Comment by cgwalters** [comment]
No, kube/crio would do the extraction.  We use `podman` on the host because during bootstrap we don't have kube up.

**Comment by cgwalters** [comment]
Actually...I think it'd be a lot simpler to change our Ignition to `podman run .../machine-config-daemon`.  Wouldn't that work?  I think that'd work.
